sdks/python/apache_beam/runners/portability/stager.py

```diff
@@ -0,0 +1,573 @@
+    for package in extra_packages:
+      if not (os.path.basename(package).endswith('.tar') or
+              os.path.basename(package).endswith('.tar.gz') or
+              os.path.basename(package).endswith('.whl') or
+              os.path.basename(package).endswith('.zip')):
+        raise RuntimeError(
+            'The --extra_package option expects a full path ending with '
+            '".tar", ".tar.gz", ".whl" or ".zip" instead of %s' % package)
+      if os.path.basename(package).endswith('.whl'):
+        logging.warning(
+            'The .whl package "%s" is provided in --extra_package. '
+            'This functionality is not officially supported. Since wheel '
+            'packages are binary distributions, this package must be '
+            'binary-compatible with the worker environment (e.g. Python 2.7 '
+            'running on an x64 Linux host).')
+
+      if not os.path.isfile(package):
+        if self._is_remote_path(package):
+          # Download remote package.
+          logging.info('Downloading extra package: %s locally before staging',
+                       package)
+          _, last_component = FileSystems.split(package)
+          local_file_path = FileSystems.join(staging_temp_dir, last_component)
+          self._download_file(package, local_file_path)
+        else:
+          raise RuntimeError(
+              'The file %s cannot be found. It was specified in the '
+              '--extra_packages command line option.' % package)
+      else:
+        local_packages.append(package)
+
+    local_packages.extend([
+        FileSystems.join(staging_temp_dir, f)
+        for f in os.listdir(staging_temp_dir)
+    ])
+
+    for package in local_packages:
+      basename = os.path.basename(package)
+      staged_path = FileSystems.join(staging_location, basename)
+      self.stage_artifact(package, staged_path)
+      resources.append(basename)
+    # Create a file containing the list of extra packages and stage it.
+    # The file is important so that in the worker the packages are installed
+    # exactly in the order specified. This approach will avoid extra PyPI
+    # requests. For example if package A depends on package B and package A
+    # is installed first then the installer will try to satisfy the
+    # dependency on B by downloading the package from PyPI. If package B is
+    # installed first this is avoided.
+    with open(os.path.join(temp_dir, EXTRA_PACKAGES_FILE), 'wt') as f:
+      for package in local_packages:
+        f.write('%s\n' % os.path.basename(package))
+    staged_path = FileSystems.join(staging_location, EXTRA_PACKAGES_FILE)
+    # Note that the caller of this function is responsible for deleting the
+    # temporary folder where all temp files are created, including this one.
+    self.stage_artifact(
+        os.path.join(temp_dir, EXTRA_PACKAGES_FILE), staged_path)
+    resources.append(EXTRA_PACKAGES_FILE)
+
+    return resources
+
+  def _get_python_executable(self):
```

Comment:
<!--new_thread; commit:fb22e20e257e1c00a618286cb0a1dd01e4071709; resolved:0-->
Make this a private function in Stager module.